### PR TITLE
feat: Live query validation in the SQL Lab UI

### LINKED ISSUE #6707: Realtime syntax validation of queries in SQL lab
Real time query validation in sql lab I think could be a very useful and powerful feature, it can bring in equivalent productivity modern IDEs bring to programming. I also noticed that query engine's like presto does have a very easy way to validate queries, prepending `explain (TYPE VALIDATE)` to a query validates the syntax of the query. Other query engine's may support similar functionalities as well. Here is some test of query validation for presto engine

## Valid query

```
presto:default> explain (TYPE VALIDATE) select * from hive.default.test_table where ds='2019-01-01' limit 10;
 Valid
-------
 true
(1 row)
```

## Invalid table name:

```presto:default> explain (TYPE VALIDATE) select * from hive.default.test_table_x where ds='2019-01-01'  limit 10;
Query 20190116_163346_02068_b85df failed: line 1:39: Table hive.default.test_table_x does not exist
explain (TYPE VALIDATE) select * from hive.default.test_table_x where ds='2019-01-01' and audit_source='presto' limit 10 
```

## Invalid column
```
presto:default> explain (TYPE VALIDATE) select * from hive.default.test_table_x where ds='2019-01-01' and col1='abc' limit 10;
Query 20190116_163535_02086_b85df failed: line 1:99: Column 'col1' cannot be resolved
explain (TYPE VALIDATE) select * from hive.default.test_table_x where ds='2019-01-01' and col1='abc' limit 10
```

More: https://prestodb.io/docs/current/sql/explain.html

Examples of such feature in other sql editors:

query-validation

### PR DESCRIPTION
**Pull request by bearcage**
### CATEGORY

Choose one

- [ ] Bug Fix
- [X] Enhancement (new features, refinement)
- [ ] Refactor
- [ ] Add tests
- [ ] Build / Development Environment
- [ ] Documentation

### SUMMARY
This builds on #7422 to build check-as-you-type sql query validation in Sql Lab. This closes #6707 too.

This adds a (debounced) call to the `validate_sql_json` API endpoint with the querytext. If no validator is configured for the selected database, per the feature flag from #7422, no validation requests will be sent from the UI. On Lyft's infra the presto validator is able to return feedback to the user (end to end) in about 2 seconds.

At present feedback is provided only through the "annotations" mechanism build in to ACE, although I'd be open to adding full text elsewhere on the page if there's interest.

This diff does not include new cypress tests covering this interaction.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

validation-med

Server error reporting:
Screen Shot 2019-05-05 at 7 19 38 PM

Query error reporting:
Screen Shot 2019-05-05 at 7 21 08 PM

### TEST PLAN
1. Start a development instance
2. Follow the instructions in INSTALLING to enable the presto validator
3. Connect to a prestodb instance
4. Try writing some queries against it!

You can verify the unhappy paths by:

* Injecting 5xx into your communications with presto
* Using a database which does not support validation (like the built-in sqlite). Note that no validation attempts will even be made in this case.

### ADDITIONAL INFORMATION
- [X] Has associated issue: #6707
- [X] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [X] Introduces new feature or API
- [ ] Removes existing feature or API

### REVIEWERS
@xtinec @DiggidyDave @khtruong @betodealmeida

### CODE REVIEW

**Comment by khtruong** [comment]
Adding @kristw and @williaster for :eyes

File: superset/assets/src/SqlLab/actions/sqlLab.js [hunk 1]
```diff
@@ -77,6 +81,26 @@ export function resetState() {
   return { type: RESET_STATE };
 }
 
+export function startQueryValidation(query) {
+  if (query == null) {
+    const msg = "programmer error: tried to validate a null query object";
+    console.error(msg);
```
Comment (khtruong) [inline superset/assets/src/SqlLab/actions/sqlLab.js]: I think we have lint rules against console statements. Maybe just throw new Error(...) instead?

File: superset/assets/src/SqlLab/actions/sqlLab.js [hunk 2]
```diff
@@ -187,6 +211,47 @@ export function runQuery(query) {
   };
 }
 
+export function validateQuery(query) {
+  return function (dispatch) {
+    if (query == null) {
+      const msg = "programmer error: tried to validate a null query object";
+      console.error(msg);
```
Comment (khtruong) [inline superset/assets/src/SqlLab/actions/sqlLab.js]: Same comment.

File: superset/assets/src/SqlLab/components/SqlEditor.jsx [hunk 3]
```diff
@@ -126,6 +133,11 @@ class SqlEditor extends React.PureComponent {
   }
   onSqlChanged(sql) {
     this.setState({ sql });
+    // Request server-side validation of the query text
+    if (this.canValidateQuery()) {
+      // NB. requestValidation is debounced
```
Comment (khtruong) [inline superset/assets/src/SqlLab/components/SqlEditor.jsx]: What is NB?

File: superset/assets/src/SqlLab/actions/sqlLab.js (hunk 1 above)
Comment (bearcage) [reply superset/assets/src/SqlLab/actions/sqlLab.js]: Those functions can't be triggered w/ a null query param right now anyway, so I thought I'd match the rest of the code in that file and drop the extra assertion

File: superset/assets/src/SqlLab/components/SqlEditor.jsx (hunk 3 above)
Comment (bearcage) [reply superset/assets/src/SqlLab/components/SqlEditor.jsx]: I never realized that was weird, but I guess it is a bit now that I think about it!

NB means "nota bene", latin for "note well" — I usually use it in comments where "NOTE:" would also work

**Review by khtruong — APPROVED** [review]

**Pull request merged**